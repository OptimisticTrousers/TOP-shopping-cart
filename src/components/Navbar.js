import logo from '../assets/optimistictrousers.jpg'
import { Link } from 'react-router-dom'
import { QuantityContext } from '../App'
import { useContext } from 'react'

export default function Navbar() {

    const handleCartQuantity = useContext(QuantityContext)

    return (
        <nav>            
            <div className="links">
                <div className="logo">
                    <Link to="/">
                        <img className="logo-image" src={logo} alt="smiling pair of pants"/>
                    </Link>
                    <h1 className="nav-title">The Optimistic Store</h1>
                    <ul>
                        <li>
                            <Link to="/">
                                <h2>HOME</h2>
                            </Link>
                        </li>
                        <li>
                            <Link to="/shop">
                                <h2>SHOP</h2>
                            </Link>
                        </li>
                    </ul>
                </div>
                    <Link to="/cart">
                        <span>{handleCartQuantity}</span>
                        <svg width="20" height="22" viewBox="0 0 20 22" fill="none" stroke="currentColor"><path d="M4 1L1 5V19C1 19.5304 1.21071 20.0391 1.58579 20.4142C1.96086 20.7893 2.46957 21 3 21H17C17.5304 21 18.0391 20.7893 18.4142 20.4142C18.7893 20.0391 19 19.5304 19 19V5L16 1H4Z" strokeWidth="1.5" strokeLinecap="round" strokeLinejoin="round"></path><path d="M1 5H19" strokeWidth="1.5" strokeLinecap="round" strokeLinejoin="round"></path><path d="M14 9C14 10.0609 13.5786 11.0783 12.8284 11.8284C12.0783 12.5786 11.0609 13 10 13C8.93913 13 7.92172 12.5786 7.17157 11.8284C6.42143 11.0783 6 10.0609 6 9" strokeWidth="1.5" strokeLinecap="round" strokeLinejoin="round"></path></svg>
                    </Link>
            </div>
        </nav>
    )
}