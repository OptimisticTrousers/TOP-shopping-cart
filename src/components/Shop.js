export default function Shop() {
    return (
        <main>
            <h1>Get Clothing!</h1>
            <p>Times are tough. Liven up with our top-notch clothing!</p>
            <div className="products">

            </div>
        </main>
    )
}