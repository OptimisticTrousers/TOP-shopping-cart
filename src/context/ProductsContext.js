import React, { useContext, useState } from "react";

const ProductsContext = React.createContext()

export function useProducts() {
    const [products] = useContext(ProductsContext)

    return products
}

export function useProductAdder() {
    const [products] = useContext(ProductsContext)

    return products
}

export function ProductsProvider({children}) {

    const [products, setProducts] = useState([])

    function addProduct(){
        
    }

    return (
        <ProductsContext.Provider value={{products, addProduct}}>
            {children}
        </ProductsContext.Provider>
    )
}