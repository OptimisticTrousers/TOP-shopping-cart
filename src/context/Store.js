import React, { useContext, useState } from "react"

const CartQuantityContext = React.createContext()
const CartContext = React.createContext()


export function useQuantity() {
    const {cartQuantity} = useContext(CartQuantityContext)

    return cartQuantity;
}

export function useAddToCart() {
    const {addToCart} = useContext(CartContext);

    return addToCart;
}

export function useCart() {
    const {cart} = useContext(CartContext)

    return cart
}

export function CartProvider({children}) {

    const [cartQuantity, setCartQuantity] = useState(0)
    const [cart, setCart] = useState([])


    const addToCart = (details) => {
        setCartQuantity(prevQuantity => prevQuantity + 1)
        setCart(prevCart => {

            const productIndex = prevCart.findIndex((product) => product.id === details.id)

            if(productIndex !== -1){
                const newCart = [...prevCart]

                newCart.splice(productIndex, 1, {...details, quantity: prevCart[productIndex].quantity + 1})

                return newCart;
            }
            return [...prevCart, {...details, quantity: 1}];
        })
    }

    return (
        <CartContext.Provider value={{cart, addToCart}}>
            <CartQuantityContext.Provider value={{cartQuantity}}>
                {children}
            </CartQuantityContext.Provider>
        </CartContext.Provider>
    )
}
