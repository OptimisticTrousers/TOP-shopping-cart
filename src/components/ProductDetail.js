import { Link, useParams } from "react-router-dom"
import { useEffect, useState } from "react"
import leftArrow from '../assets/left-arrow.svg'
import { useAddToCart } from "../context/Store"

export default function ProductDetail(){

    const addToCart = useAddToCart()

    const params = useParams()

    const [details, setDetails] = useState({})

    useEffect(() => {
        fetch(`https://fakestoreapi.com/products/${params.id}`)
            .then(res => res.json())
            .then(data => setDetails(data))
    }, [])

    return (
        <div className="detail-container">
            <div className="image-detail-view">
                <img className="image-detail" src={details.image} alt={details.title}/>
            </div>
            <div className="product-details">
                <Link to="/shop">
                    <button type="button"><img className="left-arrow" src={leftArrow} alt="left arrow" /> Back To Products</button>
                </Link>
                <h1 className="product-title">{details.title}</h1>
                <p className="product-price">${details.price}</p>
                <p className="category-detail">Category: {details.category}</p>
                <p className="description-detail">{details.description}</p>
                <button onClick={() => addToCart({...details, quantity: 0})} className="add-to-cart-button" type="button">Add to Cart</button>
            </div>
        </div>
    )
}