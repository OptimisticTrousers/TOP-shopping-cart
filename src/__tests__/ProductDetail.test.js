import {render, screen} from '@testing-library/react'
import '@testing-library/jest-dom'
import userEvent from '@testing-library/user-event'
import Shop from '../components/Shop'
import ProductDetail from '../components/ProductDetail'
import * as Store from '../context/Store'
import {BrowserRouter} from 'react-router-dom'
import Navbar from '../components/Navbar'

describe("ProductDetail", () => {

    beforeAll(() => {
        jest.mock('react-router-dom', () => ({
            ...jest.requireActual('react-router-dom'),
            useParams: () => ({
                id: 1
            }),
        }))
    })
    it("snapshot", () => {
        const {asFragment} = render(
            <BrowserRouter>
                <Store.CartProvider >
                    <ProductDetail />
                </Store.CartProvider>
            </BrowserRouter>
        )

        expect(asFragment()).toMatchSnapshot()
    })
    it("expecting to call the 'Add to Cart' function in the cart context", async () => {

        const mockAddToCart = jest.fn()
        jest.spyOn(Store, 'useAddToCart').mockReturnValue(mockAddToCart)

        render(
            <BrowserRouter>
                <Store.CartProvider >
                    <ProductDetail />
                </Store.CartProvider>
            </BrowserRouter>
        )

        const addToCartButton = await screen.findByRole('button', {name: /Add to Cart/i})

        const user = userEvent.setup()

        await user.click(addToCartButton)
        await user.click(addToCartButton)
        await user.click(addToCartButton)
        await user.click(addToCartButton)
        await user.click(addToCartButton)
        await user.click(addToCartButton)
        await user.click(addToCartButton)

        expect(mockAddToCart).toHaveBeenCalledTimes(7)
    })
    

})