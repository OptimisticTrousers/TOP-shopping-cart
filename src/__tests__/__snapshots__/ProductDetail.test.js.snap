// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`ProductDetail snapshot 1`] = `
<DocumentFragment>
  <div
    class="detail-container"
  >
    <div
      class="image-detail-view"
    >
      <img
        class="image-detail"
      />
    </div>
    <div
      class="product-details"
    >
      <a
        href="/shop"
      >
        <button
          type="button"
        >
          <img
            alt="left arrow"
            class="left-arrow"
            src="left-arrow.svg"
          />
           Back To Products
        </button>
      </a>
      <h1
        class="product-title"
      />
      <p
        class="product-price"
      >
        $
      </p>
      <p
        class="category-detail"
      >
        Category: 
      </p>
      <p
        class="description-detail"
      />
      <button
        class="add-to-cart-button"
        type="button"
      >
        Add to Cart
      </button>
    </div>
  </div>
</DocumentFragment>
`;
